### PR TITLE
Implement first-class providers for Python

### PR DESCRIPTION
**Pull request by swgillespie**
First-class providers are an explicit projection of providers themselves
into Pulumi programs. For the most post, providers are just regular
resources, but the addition of providers to the fray (and the ability of
resources to be constructed by providers in the same program) requires
some changes to the Python resource model.

A summary of the changes:

1. Added ProviderResource, a custom resource that is the base class of
all providers.
2. ResourceOptions now has 'provider' and 'providers' fields.
'provider', when passed to a custom resource, allows users to override
the provider that is used to construct a resource to an instance of a
ProviderResource. 'providers', when passed to a component resource,
allows users to override providers used to construct children of the
component resource.
3. 'protect', 'providers', and 'provider' are all now inherited from
a resource's parent if they aren't specified in the child.

This commit adds the requisite code for the above changes and, in
addition, adds a number of new tests that exercise them and related code
paths.

### CODE REVIEW

File: sdk/python/.pylintrc [hunk 1]
```diff
@@ -148,6 +148,8 @@ disable=print-statement,
         invalid-name,
         protected-access,
         too-many-arguments,
+        too-many-branches,
```
Comment (swgillespie) [inline sdk/python/.pylintrc]: As much as it pains me to ignore these lints, code in `runtime/resource.py` often trips these up and it feels like more of a pain to refactor into single-use methods then it is to just document the functions aggressively.

File: sdk/python/lib/test/langhost/asset/test_asset.py [hunk 2]
```diff
@@ -24,7 +24,7 @@ def test_asset(self):
             expected_resource_count=3)
 
     def register_resource(self, _ctx, _dry_run, ty, name, resource,
-                          _dependencies):
+                          _dependencies, _parent, _custom, _protect, _provider):
```
Comment (swgillespie) [inline sdk/python/lib/test/langhost/asset/test_asset.py]: This (and all related changes to test files) is due to a signature change of `register_resource` in `util.py`, the test harness for Python language host tests.

File: sdk/python/lib/pulumi/runtime/resource.py [hunk 3]
```diff
@@ -111,15 +147,16 @@ def do_resolve(value: Any, perform_apply: bool):
...
+        resolver = await prepare_resource(res, ty, custom, props, opts)
         log.debug(f"resource registration prepared: ty={ty}, name={name}")
         req = resource_pb2.RegisterResourceRequest(
             type=ty,
             name=name,
             parent=resolver.parent_urn,
             custom=custom,
             object=resolver.serialized_props,
-            protect=False,
+            protect=opts.protect,
```
Comment (swgillespie) [inline sdk/python/lib/pulumi/runtime/resource.py]: note: this fixes a bug where `protect` didn't work in Python... 🤦‍♂️ my bad

**Comment by swgillespie** [comment]
Fixes https://github.com/pulumi/pulumi/issues/1736, partially fixes https://github.com/pulumi/pulumi/issues/1713 (needs tfgen and kubernetes changes to expose providers in earnest)

File: sdk/python/lib/pulumi/resource.py [hunk 4]
```diff
@@ -67,8 +94,39 @@ def __init__(self,
...
+                provider = opts.provider
+                if provider is None:
+                    opts.provider = opts.parent.get_provider(t)
+                else:
+                    # If a provider was specified, add it to the providers map under this type's package so that
+                    # any children of this resource inherit its provider.
+                    type_components = t.split(":")
+                    if len(type_components) == 3:
+                        [pkg, _, _] = type_components
+                        self._providers = {**self._providers, pkg: provider}
```
Comment (pgavlin) [inline sdk/python/lib/pulumi/resource.py]: Nice that Python effectively has a spread operator :)

**Comment by pgavlin** [comment]
Though I do notice that this does not add support for FCPs to the implementation of invoke--do you plan to do that as a follow-up?

**Comment by swgillespie** [comment]
> Though I do notice that this does not add support for FCPs to the implementation of invoke--do you plan to do that as a follow-up?

Yep! That's my next PR.

**Pull request merged**